lib/SILOptimizer/Utils/Generics.cpp

```diff
@@ -166,6 +174,164 @@ static bool isTypeTooComplex(Type t) {
   return TypeWidth >= TypeWidthThreshold || TypeDepth >= TypeDepthThreshold;
 }
 
+namespace {
+
+/// A helper class used to check whether one type is structurally contained
+/// the other type either completely or partially.
+class TypeComparator : public TypeMatcher<TypeComparator> {
+  bool IsContained = false;
+
+public:
+  bool isEqual(CanType T1, CanType T2) { return T1 == T2; }
+  /// Check whether the type T1 is different from T2 and contained in the type
+  /// T2.
+  bool isStrictlyContainedIn(CanType T1, CanType T2) {
+    if (isEqual(T1, T2))
+      return false;
+    return T2.findIf([&T1, this](Type T) -> bool {
+      return isEqual(T->getCanonicalType(), T1);
+    });
+  }
+
+  /// Check whether the type T1 is strictly or partially contained in the type
+  /// T2.
+  /// Partially contained means that if you drop the common structural "prefix"
+  /// of T1 and T2 and get T1' and T2' then T1' is strictly contained in T2'.
+  bool isPartiallyContainedIn(CanType T1, CanType T2) {
+    if (isStrictlyContainedIn(T1, T2))
+      return true;
+    match(T1, T2);
+    return IsContained;
+  }
+
+  /// This method is invoked aftre skipping a common prefix of two types,
+  /// when a structural difference is found.
+  bool mismatch(TypeBase *firstType, TypeBase *secondType,
+                Type sugaredFirstType) {
+    auto firstCanType = firstType->getCanonicalType();
+    auto secondCanType = secondType->getCanonicalType();
+    if (isEqual(firstCanType, secondCanType))
+      return false;
+    if (isStrictlyContainedIn(firstCanType, secondCanType)) {
+      IsContained = true;
+      return false;
+    }
+    return false;
+  }
+};
+
+} // anonymous namespace
+
+/// Checks if a second substitution list is an expanded version of
+/// the first substitution list.
+/// This is the case if at least one of the substitution type in Subs2 is
+/// "bigger" than the corresponding substitution type in Subs1.
+/// Type T2 is "smaller" than type T1 if T2 is structurally contained in T1.
+static bool growingSubstitutions(SubstitutionList Subs1,
+                                 SubstitutionList Subs2) {
+  assert(Subs1.size() == Subs2.size());
+  TypeComparator TypeCmp;
+  // Perform component-wise comparisions for substitutions.
+  for (unsigned idx = 0, e = Subs1.size(); idx < e; ++idx) {
+    auto Type1 = Subs1[idx].getReplacement()->getCanonicalType();
+    auto Type2 = Subs2[idx].getReplacement()->getCanonicalType();
+    // Replacement types should be concrete.
+    assert(!Type1->hasArchetype());
+    assert(!Type2->hasArchetype());
+    // If types are the same, the substitution type does not grow.
+    if (TypeCmp.isEqual(Type2, Type1))
+      continue;
+    // If the new substitution type is getting smaller, the
+    // substitution type does not grow.
+    if (TypeCmp.isStrictlyContainedIn(Type2, Type1))
```

Comment:
Could you expand on why this is strict? It seems like one could similarly have the interior of the types shrink, e.g. `Outer<Step<Step<Start>>>` -> `Outer<Step<Start>>` -> `Outer<Start>`?